### PR TITLE
Add CT Clamp component

### PR DESCRIPTION
**Pull request by jesserockz**
## Description:
Adds sensor code for CT (Current Transformer) Clamps.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#256

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ct_clamp/ct_clamp_sensor.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  ESP_LOGCONFIG(TAG, "  Sample Size: %u", this->sample_size_);
+  ESP_LOGCONFIG(TAG, "  Supply Voltage: %0.2fV", this->supply_voltage_);
+  LOG_UPDATE_INTERVAL(this);
+}
+
+float CTClampSensor::get_setup_priority() const { return setup_priority::DATA; }
+
+void CTClampSensor::update() {
+  for (unsigned int n = 0; n < this->sample_size_; n++) {
+    sample_i_ = analogRead(this->pin_);
```
Comment (OttoWinter) [inline esphome/components/ct_clamp/ct_clamp_sensor.cpp]: This references `analogRead` directly and is thus not really platform-independent (and other voltage sensors like ADS1115 can't be used with this component).

Instead, make this component accept a `voltage_sampler::VoltageSampler` instance and use that to get the voltage readings.

This loop would then look more like 

```cpp
float sample = this->voltage_sensor_->sample();
```

File: esphome/components/ct_clamp/ct_clamp_sensor.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  ESP_LOGCONFIG(TAG, "  Calibration: %u", this->calibration_);
+  ESP_LOGCONFIG(TAG, "  Sample Size: %u", this->sample_size_);
+  ESP_LOGCONFIG(TAG, "  Supply Voltage: %0.2fV", this->supply_voltage_);
+  LOG_UPDATE_INTERVAL(this);
+}
+
+float CTClampSensor::get_setup_priority() const { return setup_priority::DATA; }
+
+void CTClampSensor::update() {
+  for (unsigned int n = 0; n < this->sample_size_; n++) {
```
Comment (OttoWinter) [inline esphome/components/ct_clamp/ct_clamp_sensor.cpp]: I don't see a `delay` (or similar call) in this loop. Could you explain how this calculation works (or refer to some formulas)? This looks a lot like code that was written for arduino etc for me, but might not work well on an ESP with much higher clock rates.

File: esphome/components/ct_clamp/ct_clamp_sensor.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#include "esphome/components/ct_clamp/ct_clamp_sensor.h"
```
Comment (OttoWinter) [inline esphome/components/ct_clamp/ct_clamp_sensor.cpp]:
```suggestion
#include "ct_clamp_sensor.h"
```

File: esphome/components/ct_clamp/ct_clamp_sensor.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+#endif
+
+ protected:
+  uint8_t pin_;
+  uint32_t calibration_;
+  uint32_t sample_size_;
+  double supply_voltage_;
+
+  int sample_i_;
+  double irms_, filtered_i_, offset_i_, sum_i_, sq_i_;
```
Comment (OttoWinter) [inline esphome/components/ct_clamp/ct_clamp_sensor.h]: Why are these instance attributes?

These values only seem to be used in the update function, so I don't see a need for them to be declared here?

Also, don't use double unless double accuracy is really required. ESPHome uses floats everywhere as that saves code space and double is rarely actually needed.

File: esphome/components/ct_clamp/ct_clamp_sensor.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  double i_ratio = this->calibration_ * ((supply_voltage_) / (ADC_COUNTS));
+  irms_ = i_ratio * sqrt(sum_i_ / this->sample_size_);
+
+  // Reset accumulators
+  sum_i_ = 0;
+
+  ESP_LOGD(TAG, "'%s'", this->get_name().c_str());
+  ESP_LOGD(TAG, "   Amps=%.2fA", irms_);
```
Comment (OttoWinter) [inline esphome/components/ct_clamp/ct_clamp_sensor.cpp]: Why two log lines?

```suggestion
  ESP_LOGD(TAG, "'%s' - Amps: %.2fA", this->name_.c_str(), irms_);
```

File: esphome/components/ct_clamp/ct_clamp_sensor.cpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  ESP_LOGCONFIG(TAG, "Setting up CT Clamp '%s'...", this->get_name().c_str());
+  GPIOPin(this->pin_, INPUT).setup();
+
+  offset_i_ = ADC_COUNTS >> 1;
+}
+
+void CTClampSensor::dump_config() {
+  LOG_SENSOR("", "CT Clamp Sensor", this);
+  ESP_LOGCONFIG(TAG, "  Pin: %u", this->pin_);
+  ESP_LOGCONFIG(TAG, "  Calibration: %u", this->calibration_);
```
Comment (OttoWinter) [inline esphome/components/ct_clamp/ct_clamp_sensor.cpp]: I don't think this is necessary - calibration is a thing sensor filters should do: multiply, calibrate_linear etc.

**Comment by jesserockz** [comment]
Hi @OttoWinter 
I took the bulk of this code from [EMonLib](https://github.com/openenergymonitor/EmonLib) and made sure it works and is accurate enough on the ESP (Using an external power monitor along side this to show the same results.)

I will make those suggested changes and try to tidy up the variables.

Cheers,
Jesse

File: esphome/components/ct_clamp/ct_clamp_sensor.cpp (hunk 1 above)
Comment (jesserockz) [reply esphome/components/ct_clamp/ct_clamp_sensor.cpp]: Do you have any more examples of this? I cannot find any components in dev that do this already? 
They all seem to implement a `sample` function and then just have `analogRead` inside that?

Comment (OttoWinter) [reply esphome/components/ct_clamp/ct_clamp_sensor.cpp]: I think there are only VoltageSampler providers out there right now - not consumers.

The process is quite simple though:

In .py:
```python
AUTO_LOAD = ['voltage_sampler']

CONFIG_SCHEMA = ...
  cv.Required(CONF_SENSOR): cv.use_id(voltage_sampler.VoltageSampler),

def to_code(config):
   ...
   sens = yield cg.get_variable(config[CONF_SENSOR])
   cg.add(var.set_source(sens))
```

```cpp
class ... {
  public:
    void set_source(voltage_sampler::VoltageSampler *source) { source_ = source; }
  protected:
    voltage_sampler::VoltageSampler *source_;
};

// in update
float value = this->source_->sample();
```

Comment (jesserockz) [reply esphome/components/ct_clamp/ct_clamp_sensor.cpp]: Ah, I get it now. I was thinking in a different way. I will give that a shot. Thanks

**Comment by jesserockz** [comment]
@OttoWinter I have made some changes so that it uses a VoltageSampler sensor to read the analog in and also moved some of the variables to be filters.
My config:

```yaml

sensor:
  - platform: adc
    id: analog
    pin: A0

  - platform: ct_clamp
    name: Measured Current
    sensor: analog
    sample_size: 1637 # 10 cycles at 50Hz frequency on ESP8266
    update_interval: 500ms
    filters:
      - multiply: 30 # Calibration for resister
      - multiply: 3.26 # Calibration for D1 Mini voltage converter
```

**Review by OttoWinter — COMMENTED** [review]
Hi!

Thanks for the quick update of the component!

There's one small problem with it though when I looked at it again - the update() method of course needs to sample over a couple of AC phases. But the problem with that is that it slows down ESPHome's main loop - usually slow downs of up to around 20-30ms are ok but here it's more in the 100ms range.

I've updated the code to use the loop() method in ESPHome and be more asynchronous. Now the update() method only starts a "sampling" process and the loop() method does the actual sampling. At the end of this sampling period (defined by `sample_duration`), a timeout triggers and publishes the result.

Can you confirm the updated code is working? I've just ordered a CT clamp myself, but it'll take some time to arrive, and I hope to get this into 1.13 :)  (that also requires docs though)

**Comment by jesserockz** [comment]
Wow, I saw the changes and I am pushing to my device now to test them out. 
Will let you know how it goes.

**Comment by jesserockz** [comment]
Updated and still giving me the same values back from my usage. 👍 
I did create the docs at esphome/esphome-docs#256, will just need to update them with the configuration changes.

**Comment by OttoWinter** [comment]
Ah sorry, missed that PR.

I'll tag this PR for 1.13 then, great!

**Comment by digiblur** [comment]
@jesserockz I played around with this with my currently built sensor.  Couple things the docs were missing the liner filter part so it wouldn't compile and then I'm not getting consistent readings for some reason like I was with emonlib.  Curious of what setup you are using for testing?

**Comment by jesserockz** [comment]
Hi @digiblur,
I will check the docs for that.

I am using a `CTL-10-CLS` Clamp wired to my D1 Mini A0 pin via the required circuitry with burden resistor and resistor-based voltage divider.
I have the clamp attached to my main incoming "live" line for the whole house (single phase AC in NZ) and I get readings consistent with my "Smart" meter.

**Comment by digiblur** [comment]
Thanks for the quick comment.  I'll give it shot with a bigger load and see how it works.  Using 30A clamp on about a 1 amp draw, looks fine with arduino emonlib code so I know the hardware is right.  I'll toy around with it some more.  

Thanks for your efforts on this PR btw!

**Pull request merged**